src/site/topics/bags/bag-adt.rst

```diff
@@ -1,214 +1,270 @@
+    A representation of an arbitrary bag containing seven elements.
+
+
+* The high-level idea of a bag shouldn't really be thought of as having an ordering
 
-* What things do we want to do with our new data structure?
-    * Add things
-    * Remove things
+    * The underlying implementation may be with some linear container
+    * But the idea is that the information in the bag has no meaningful order
+
+
+* Like the other collections, there needs to be a way to
+
+    * Add
+    * Remove
     * Get the size
     * Check if it's empty
 
-* Adding to a Bag
-    * How should we add things?
-    * Where should they go?
 
-* Removing from a Bag
-    * How should we remove things?
-    * Where should they be removed from?
+* But given the more general definition of the bag, how exactly these should be done is not obvious
 
-* Other things me may want to do?
-    * Take a minute and think about potential ideas
+    * Adding to a Bag
 
+        * How should elements be added?
+        * Where should they go?
 
-Sorted Bags
------------
 
-* If we want a specific kind of bag that keeps our elements sorted based on some ordering, adding will need to be done carefully
+    * Removing from a Bag
+
+        * How should elements be removed?
+        * Where should they be removed from?
+
+
+    * Other potential functionality?
+
+        * Access a specific element?
+        * Check if a given element exists?
+        * ...
+
+
+Sorted Bag
+----------
+
+* If the elements must be sorted, then how they are added and removed must be done carefully
+
     * Unlike the general bag, this specific kind of bag will have an ordering
 
+
 * The ordering of the elements will depend on some characteristic of the contents of the bag
+
     * Numbers in ascending order
     * Strings in alphabetical order
 
+
 * The elements themselves are what determine the ordering
 
-.. image:: bag_sorted_bag0.png
-   :width: 500 px
-   :align: center
+.. figure:: bag_sorted_bag0.png
+    :width: 500 px
+    :align: center
+
+    Example sorted bag of seven integers. The integer 58 is to be added to the collection.
+
+
+* There is only one way to add the element to the sorted bag
+
+    * The element must be added such that the sorted property of the bag is preserved
+
+
+.. figure:: bag_sorted_bag1.png
+    :width: 500 px
+    :align: center
+
+    Example of inserting the integer ``58`` into the only valid spot in the sorted bag. Notice that the integer ``67``
+    is the first element within the sorted bag that is larger than the element being added.
 
-* There really is only one way to add elements to this sorted bag
-    * The elements **must** go where it belongs
-* We cannot specify where we add it because that may break our sorted ordering
 
-.. image:: bag_sorted_bag1.png
-   :width: 500 px
-   :align: center
+* Should there be restrictions on how elements are removed?
 
-* Should there be rules restricting removing from the bag?
 
 
 Indexed Bags
 ------------
```

Comment:
Other implementations may include `MapBag` (so `HashMapBag`, `TreeMapBag`), where the bag itself is backed by a `Map<T, Integer>`. Iteration for these gets more complex though, since you need to repeat elements that aren't actually there.